[WGSL] Explicit type conversion should not check for overflow

tadeuzagallo opened the pull request:
#### f0f4573a7ded9af3e7e471119cb17cee0f25a0d1
<pre>
[WGSL] Explicit type conversion should not check for overflow
<a href="https://bugs.webkit.org/show_bug.cgi?id=265688">https://bugs.webkit.org/show_bug.cgi?id=265688</a>
<a href="https://rdar.apple.com/119044778">rdar://119044778</a>

Reviewed by Mike Wyrzykowski.

When converting values to a different type (e.g. u32(i32)), we should not perform
the same overflow checks as when a type is implicitly converted (e.g. from AbstractInt
to i32).

* Source/WebGPU/WGSL/ConstantFunctions.h:
(WGSL::convertValue):
(WGSL::constantConstructor):
(WGSL::constantVector):
(WGSL::constantMatrix):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::chooseOverload):
(WGSL::TypeChecker::convertValue):
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: <a href="https://commits.webkit.org/271499@main">https://commits.webkit.org/271499@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f04ddc7ad218ece40ec133c9f17a71481f11690

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28513 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29899 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31040 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25967 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9338 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4526 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28782 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24536 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5171 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5343 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31725 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26169 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25978 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31568 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5253 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3429 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29339 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/notification (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6855 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6845 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5710 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5773 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -1419,31 +1419,24 @@ const Type* TypeChecker::chooseOverload(const char* kind, AST::Expression& expre
             isConstructor = true;
```
looks like this needs to be deleted?

tadeuzagallo committed:
[WGSL] Explicit type conversion should not check for overflow
https://bugs.webkit.org/show_bug.cgi?id=265688
rdar://119044778

Reviewed by Mike Wyrzykowski.

When converting values to a different type (e.g. u32(i32)), we should not perform
the same overflow checks as when a type is implicitly converted (e.g. from AbstractInt
to i32).

* Source/WebGPU/WGSL/ConstantFunctions.h:
(WGSL::convertValue):
(WGSL::constantConstructor):
(WGSL::constantVector):
(WGSL::constantMatrix):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::chooseOverload):
(WGSL::TypeChecker::convertValue):
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: https://commits.webkit.org/271499@main
